### ISSUE TITLE
Compilation warnings with gcc12.1

### ISSUE THREAD

**Issue by tkittel** [mctools/ncrystal#100]
While working on #99, I spotted several compilation warnings when building in the context of `conda build`. This is most likely due to the fact that this builds with gcc12.1:
```
[  3%] Building CXX object CMakeFiles/NCrystal.dir/ncrystal_core/src/NCAtomData.cc.o
In file included from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/src/NCAtomData.cc:24:
/home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/internal/NCIter.hh:50:33: warning: 'template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator' is deprecated [-Wdeprecated-declarations]
   50 |   class enumerate_iter_t : std::iterator<std::forward_iterator_tag, typename std::iterator_traits<TIter>::value_type> {
      |                                 ^~~~~~~~
In file included from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/_build_env/x86_64-conda-linux-gnu/include/c++/12.1.0/bits/stl_algobase.h:65,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/_build_env/x86_64-conda-linux-gnu/include/c++/12.1.0/bits/specfun.h:45,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/_build_env/x86_64-conda-linux-gnu/include/c++/12.1.0/cmath:1935,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCDefs.hh:29,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:25,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCTypes.hh:25,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCAtomData.hh:24,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/src/NCAtomData.cc:21:
/home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/_build_env/x86_64-conda-linux-gnu/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
[  4%] Building CXX object CMakeFiles/NCrystal.dir/ncrystal_core/src/NCAtomUtils.cc.o
In file included from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/internal/NCAtomUtils.hh:24,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/src/NCAtomUtils.cc:21:
In static member function 'static void NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::clear(NCrystal::SmallVector<TValue, NSMALL, MODE>*) [with TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]',
    inlined from 'void NCrystal::SmallVector<TValue, NSMALL, MODE>::clear() [with TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:721:16,
    inlined from 'static void NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::adoptHeap(NCrystal::SmallVector<TValue, NSMALL, MODE>*, DetachedHeap&) [with TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:333:18,
    inlined from 'static void NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::resizeLargeCapacity(NCrystal::SmallVector<TValue, NSMALL, MODE>*, NCrystal::SmallVector<TValue, NSMALL, MODE>::size_type) [with TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:367:16,
    inlined from 'static TValue& NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::grow_and_emplace_back(NCrystal::SmallVector<TValue, NSMALL, MODE>*, Args&& ...) [with Args = {NCrystal::StrView}; TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:442:28,
    inlined from 'static TValue& NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::emplace_back(NCrystal::SmallVector<TValue, NSMALL, MODE>*, Args&& ...) [with Args = {NCrystal::StrView}; TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:460:48,
    inlined from 'TValue& NCrystal::SmallVector<TValue, NSMALL, MODE>::emplace_back(Args&& ...) [with Args = {NCrystal::StrView}; TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:715:30,
    inlined from 'static TValue& NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::grow_and_emplace_back(NCrystal::SmallVector<TValue, NSMALL, MODE>*, Args&& ...) [with Args = {NCrystal::StrView}; TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:443:34,
    inlined from 'static TValue& NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::emplace_back(NCrystal::SmallVector<TValue, NSMALL, MODE>*, Args&& ...) [with Args = {NCrystal::StrView}; TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:460:48:
/home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:393:18: warning: 'void free(void*)' called on pointer 'result' with nonzero offset [1, 9223372036854775807] [-Wfree-nonheap-object]
  393 |         std::free(THIS->m_data.large.data);
      |         ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCDefs.hh:55,
                 from /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:25:
In function 'void* NCrystal::alignedAlloc(std::size_t, std::size_t)',
    inlined from 'TValue* NCrystal::alignedAlloc(std::size_t) [with TValue = StrView]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCMem.hh:330:45,
    inlined from 'static NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::DetachedHeap NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::createNewDetachedHeap(NCrystal::SmallVector<TValue, NSMALL, MODE>::size_type) [with TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:343:40,
    inlined from 'static void NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::resizeLargeCapacity(NCrystal::SmallVector<TValue, NSMALL, MODE>*, NCrystal::SmallVector<TValue, NSMALL, MODE>::size_type) [with TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:364:12,
    inlined from 'static TValue& NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::grow_and_emplace_back(NCrystal::SmallVector<TValue, NSMALL, MODE>*, Args&& ...) [with Args = {NCrystal::StrView}; TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:442:28,
    inlined from 'static TValue& NCrystal::SmallVector<TValue, NSMALL, MODE>::Impl::emplace_back(NCrystal::SmallVector<TValue, NSMALL, MODE>*, Args&& ...) [with Args = {NCrystal::StrView}; TValue = NCrystal::StrView; long unsigned int NSMALL = 8; NCrystal::SVMode MODE = NCrystal::SVMode::FASTACCESS]' at /home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCSmallVector.hh:460:48:
/home/tkittel/scr/miniforge_install/miniforge_installed/envs/standalone/conda-bld/ncrystal_1661156287175/work/src/ncrystal_core/include/NCrystal/NCMem.hh:318:34: note: returned from 'void* malloc(size_t)'
  318 |       void * result = std::malloc(size);
```

**Comment by MilanKlausz** [comment]
Similar compillation warning is raised by AppleClang 13.1.6.13160021 (Xcode_13.4.1) - warning about the depreciated std::iterator in NCrystalRel/inc/NCrystalRel/internal_NCIter.hh:50:33 - when testing the DGCode Framework with a macos-12 GitHub runner.

```
  Building NCrystalRel/libsrc/NCVDOSEval_rel.o
In file included from /Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pkgs/NCrystalRel/libsrc/NCVDOSEval_rel.cc:23:
/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pc/NCrystalRel/inc/NCrystalRel/internal_NCIter.hh:50:33: error: 'iterator<std::forward_iterator_tag, double>' is deprecated [-Werror,-Wdeprecated-declarations]
  class enumerate_iter_t : std::iterator<std::forward_iterator_tag, typename std::iterator_traits<TIter>::value_type> {
                                ^
/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pkgs/NCrystalRel/libsrc/NCVDOSEval_rel.cc:322:20: note: in instantiation of template class 'NCrystalRel::enumerate_iter_t<std::__wrap_iter<const double *>>' requested here
      for ( auto e : enumerate(grid) ) {
                   ^
  Building NCrystalRel/libsrc/NCVDOSGn_rel.o
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__iterator/iterator.h:27:29: note: 'iterator<std::forward_iterator_tag, double>' has been explicitly marked deprecated here
struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 iterator
                            ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1066:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1043:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                               ^
1 error generated.
make: *** [/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pc/NCrystalRel/objs/libsrc/NCVDOSEval_rel.o] Error 1
make: *** Waiting for unfinished jobs....
In file included from /Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pkgs/NCrystalRel/libsrc/NCVDOSGn_rel.cc:25:
/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pc/NCrystalRel/inc/NCrystalRel/internal_NCIter.hh:50:33: error: 'iterator<std::forward_iterator_tag, double>' is deprecated [-Werror,-Wdeprecated-declarations]
  class enumerate_iter_t : std::iterator<std::forward_iterator_tag, typename std::iterator_traits<TIter>::value_type> {
                                ^
/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pkgs/NCrystalRel/libsrc/NCVDOSGn_rel.cc:149:14: note: in instantiation of template class 'NCrystalRel::enumerate_iter_t<std::__wrap_iter<double *>>' requested here
  for (auto e: enumerate(egrid) ) {
             ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__iterator/iterator.h:27:29: note: 'iterator<std::forward_iterator_tag, double>' has been explicitly marked deprecated here
struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 iterator
                            ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1066:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1043:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                               ^
In file included from /Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pkgs/NCrystalRel/libsrc/NCVDOSGn_rel.cc:25:
/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pc/NCrystalRel/inc/NCrystalRel/internal_NCIter.hh:50:33: error: 'iterator<std::forward_iterator_tag, double>' is deprecated [-Werror,-Wdeprecated-declarations]
  class enumerate_iter_t : std::iterator<std::forward_iterator_tag, typename std::iterator_traits<TIter>::value_type> {
                                ^
/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pkgs/NCrystalRel/libsrc/NCVDOSGn_rel.cc:225:16: note: in instantiation of template class 'NCrystalRel::enumerate_iter_t<std::__wrap_iter<const double *>>' requested here
  for ( auto e :  enumerate(spec) ) {
               ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__iterator/iterator.h:27:29: note: 'iterator<std::forward_iterator_tag, double>' has been explicitly marked deprecated here
struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 iterator
                            ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1066:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/Applications/Xcode_13.4.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1043:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                               ^
2 errors generated.
make: *** [/Users/runner/work/dgcode_actions/dgcode_actions/projects/.bld/pc/NCrystalRel/objs/libsrc/NCVDOSGn_rel.o] Error 1
ERROR: Build problems encountered
```

**Comment by tkittel** [comment]
The `std::iterator` issue was solved in release 3.3.1. I am not 100% sure that there we no other issues above, so leaving open this issue for a while.

**Comment by tkittel** [comment]
Closing this now.
